sharedbuffers/mapped_struct.py

```diff
@@ -1425,21 +1576,52 @@ def proxied_list_richcmp(a, b, op):
     else:
         raise NotImplementedError
 
+def smallest_diff_key(A, B):
+    """return the smallest key adiff in A such that adiff not in B or A[adiff] != B[bdiff]"""
+    diff_keys = [k for k in A if k not in B or A[k] != B[k]]
+    return min(diff_keys)
+
+def dict_cmp_py3(A, B):
+    if len(A) != len(B):
+        return cmp(len(A), len(B))
+    try:
+        adiff = smallest_diff_key(A, B)
+    except ValueError:
+        # No difference.
+        return 0
+    bdiff = smallest_diff_key(B, A)
+    if adiff != bdiff:
+        return cmp(adiff, bdiff)
+    return cmp(A[adiff], B[bdiff])
+
+def dict_cmp(a, b):
+    if python2:
+        return cmp(a,b)
+    else:
+        return dict_cmp_py3(a,b)
+
 #@cython.ccall
 #@cython.returns(int)
 def proxied_list_cmp(a, b):
 
     alen = len(a)
     blen = len(b)
 
-    for i in xrange(min(alen, blen)):
+    for i in range(min(alen, blen)):
         selfe = a[i]
         othere = b[i]
 
-        if selfe < othere:
-            return -1
-        elif selfe > othere:
-            return 1
+        if isinstance(selfe, dict):
+            r = dict_cmp(selfe, othere)
+            if r:
+                return r
```

Comment:
Why this?

If it's to support comparing lists that contain dicts, py3 doesn't support it, so it's fine if our proxied list doesn't either:

``` pycon
>>> [{3:4}] < [{}]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'
```

Reply:
Ok... but this breaks `tests.mapped_struct.DictContainerPackingTest`
(also `None` vs `None` works in py2 but not in py3, there's a test that uses this)

Reply:
Maybe we can skip those tests for py3 (how?)

Reply:
> Ok... but this breaks `tests.mapped_struct.DictContainerPackingTest`

Well, the test should not be testing something that's illegal in py3.

>  (also `None` vs `None` works in py2 but not in py3, there's a test that uses this)

Not sure what you mean by that. I can compare None-containing sutff quite well in py3. Do you have a case at hand?

> Maybe we can skip those tests for py3 (how?)

We could. At least those that should not be passing in py3, but lets check that it is actually the case before we start skipping tests.

Reply:
Yep, that's what I did. Sooo, what about the dict stuff?

Reply:
I see the problem.

`proxied_list_richcmp` uses `proxied_list_cmp` to do the comparison, but dicts and frozensets don't have a total order in py3, and it raises an exception rather than invent a total order the way py2 does.

I mantain that we shouldn't be implementing comparisons that aren't valid in py3, so I think the best option is to implement a `proxied_list_eq` specializing in equality comparisons, to avoid depending on the inequalities for tuple values.

Then `proxied_list_richcmp` can use the eq one for `==` and `!=` checks, and `proxied_list_cmp` for the rest.